src/main/java/com/yelp/nrtsearch/server/luceneserver/field/IdFieldDef.java

```diff
@@ -0,0 +1,145 @@
+import java.io.IOException;
+import java.util.List;
+import org.apache.lucene.document.BinaryDocValuesField;
+import org.apache.lucene.document.Document;
+import org.apache.lucene.document.FieldType;
+import org.apache.lucene.index.BinaryDocValues;
+import org.apache.lucene.index.DocValues;
+import org.apache.lucene.index.DocValuesType;
+import org.apache.lucene.index.IndexOptions;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.util.BytesRef;
+
+/** Field class for defining '_ID' fields which are used to update documents */
+public class IdFieldDef extends IndexableFieldDef {
+
+  protected IdFieldDef(String name, Field requestField) {
+    super(name, requestField);
+  }
+
+  /**
+   * We use _ID fields as strings to store and update documents. Hence we fail if the field requests
+   * for properties that are not applicable for this use
+   *
+   * @param requestField field properties to validate
+   */
+  protected void validateRequest(Field requestField) {
+    if (requestField.getMultiValued() || requestField.getTokenize()) {
+      throw new IllegalArgumentException(
+          String.format(
+              "field: %s cannot have multivalued fields or tokenization as it's an _ID field",
+              requestField.getName()));
+    }
+    if (!requestField.getStore()) {
+      throw new IllegalArgumentException(
+          String.format(
+              "field: %s is an _ID field and should have store=true", requestField.getName()));
+    }
+  }
+
+  /**
+   * _ID fields should always have indexing turned on so that they can be retrieved for updates
+   * Also, these fields cannot be tokenized since we would like to use the values as it is
+   *
+   * @param fieldType type that needs search properties set
+   * @param requestField field from request
+   */
+  protected void setSearchProperties(FieldType fieldType, Field requestField) {
+    fieldType.setIndexOptions(IndexOptions.DOCS);
+    fieldType.setOmitNorms(true);
+    fieldType.setTokenized(false);
+  }
+
+  /**
+   * Store the docvalues if it's requested and store the string value in the document
+   *
+   * @param document lucene document to be added to the index
+   * @param fieldValues list of String encoded field values
+   * @param facetHierarchyPaths list of list of String encoded paths for each field value be
+   */
+  @Override
+  public void parseDocumentField(
+      Document document, List<String> fieldValues, List<List<String>> facetHierarchyPaths) {
+    if (fieldValues.size() > 1) {
+      throw new IllegalArgumentException("Cannot index multiple values into _id fields");
+    }
+    String fieldStr = fieldValues.get(0);
+    if (hasDocValues()) {
+      BytesRef stringBytes = new BytesRef(fieldStr);
+      document.add(new BinaryDocValuesField(getName(), stringBytes));
+    }
+    if (isStored()) {
+      document.add(new FieldWithData(getName(), fieldType, fieldStr));
+    }
```

Comment:
I don't think it is required for the value to be stored. Stored is a property of FieldType and will not affect the other indexing options.
You should be able to always add FieldWithData to the document. Even if stored was false in the field message, the value should still be indexed. I think indexing is enough to make updateDocument work.

That being said, it would probably be good to force the field to be retrievable somehow (doc value or stored).

Reply:
That's true. I changed the check to be either store=true or docvalue=true in `validateRequest`